torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -137,6 +137,81 @@ const std::exception& ProcessGroupMPI::WorkMPI::exception() const {
   }
 }
 
+ProcessGroupMPI::AsyncWork::AsyncWork(
+    at::Tensor tensor,
+    MPI_Request request,
+    int* srcRank)
+    : tensor_(std::move(tensor)),
+      request_(request),
+      srcRank_(srcRank) {
+  memset(&status_, 0, sizeof(status_));
+}
+
+ProcessGroupMPI::AsyncWork::~AsyncWork() {
+  if (request_ != MPI_REQUEST_NULL) {
+    throw std::runtime_error(
+        "Invalid call to AsyncWork::~AsyncWork before work has completed");
+  }
+}
+
+bool ProcessGroupMPI::AsyncWork::isCompleted() {
+  if (request_ == MPI_REQUEST_NULL) {
+    return true;
+  }
+
+  std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+  int flag = 0;
+  MPI_CHECK(MPI_Test(&request_, &flag, &status_));
+  if (request_ != MPI_REQUEST_NULL) {
+    return false;
+  }
+
+  // request_ == MPI_REQUEST_NULL; the work has completed
+  if (srcRank_ != nullptr) {
+    *srcRank_ = status_.MPI_SOURCE;
+  }
+
+  return true;
+}
+
+bool ProcessGroupMPI::AsyncWork::isSuccess() const {
+  if (request_ != MPI_REQUEST_NULL) {
+    throw std::runtime_error(
+        "Invalid call to AsyncWork::isSuccess before work has completed");
+  }
+
+  return status_.MPI_ERROR == MPI_SUCCESS;
+}
+
+void ProcessGroupMPI::AsyncWork::synchronize() {}
+
+bool ProcessGroupMPI::AsyncWork::wait() {
+  if (request_ == MPI_REQUEST_NULL) {
+    return true;
+  }
+
+  std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+  MPI_CHECK(MPI_Wait(&request_, &status_));
+  if (srcRank_ != nullptr && status_.MPI_ERROR == MPI_SUCCESS) {
+    *srcRank_ = status_.MPI_SOURCE;
+  }
+
+  return status_.MPI_ERROR == MPI_SUCCESS;
+}
+
+const std::exception& ProcessGroupMPI::AsyncWork::exception() const {
+  if (request_ != MPI_REQUEST_NULL) {
+    throw std::runtime_error(
+        "Invalid call to AsyncWork::exception before work has completed");
+  }
+
+  std::cout << "exception, mpi error: " << status_.MPI_ERROR << std::endl;
+  std::array<char, MPI_MAX_ERROR_STRING> buf;
```

Comment:
Can't this be a std::string

Reply:
Not possible: need a non-const char* of size MPI_MAX_ERROR_STRING and std::string only gives const pointers to its data (either via c_str() or data()).